### DISCUSSION TITLE
Make AI dumber for Instant battles option and AUTO mode

### DISCUSSION THREAD

**Discussion by ihhub** [ihhub/fheroes2#6260] [Ideas]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

I noticed that AI has become too smart and sometimes wins battles with even better results that an average human player. Instant battles are designed to help players in easy battles but not in complicated ones. The current state of the options leads to a situation that players don't even play battles and only focus on Adventure Map strategy. This shouldn't (my own opinion) happen in general as players suppose to play the full game.

### Describe the possible solution

Just make AI decisions in battle dumber. One way to do this is to always make AI to attack without defence actions. In this case AI might loose more army without trying to defend their units.

### Additional info

This topic open for discussions but I suggest to explain an easy to implement solution from development point of view.

**Comment by oleg-derevenetz** [comment]
I periodically watch HotA streams from pro gamers and even there, it is not so rare for AI to cope better than this pro player (he decides that he can do better, replays the battle and ends up with the worst losses), let alone an ordinary player. And no one raises the question about making the AI dumber :)

**Comment by Branikolog** [comment]
I agree with @ihhub . Autobattle AI shouldn't perform his best. I think, we shouldn't make him the dumbest player ever, but certainly now he plays better, than any novice player. So, how would new players study heroes battles then? :) Instant battle option is designed to work only for the cases where player's army is significantly stronger, than opponent's one.
Even in H3 autobattle AI behaves straightly: no waiting, no maneuvering. I think we should allow our instant AI to play similar. Without picking proper target (should attack the nearest one), always rushing ahead etc.

**Comment by Alucard648** [comment]
I wish that dumber insta-battles AI should be applicable only on higher difficulties. Maybe Rook or higher.

**Comment by ihhub** [comment]
Actually, @Alucard648 has a point. On higher difficulties it is reasonable to make AI dumber.

**Comment by zenseii** [comment]
Hi, everyone.

Am I misunderstanding or are we saying to make the AI progressively dumber the higher the difficulty is set to?

This seems contrary to what the opening message states. I think in a perfect world the instant battles would play exactly as you would. Since this is impossible I think the instant battles for easy difficulty should perform worse than for impossible difficulty. This should avoid novice players abusing it and not gaining any combat skills.

**Comment by Alucard648** [comment]
> This seems contrary to what the opening message states. I think in a perfect world the instant battles would play exactly as you would. Since this is impossible I think the instant battles for easy difficulty should perform worse than for impossible difficulty. This should avoid novice players abusing it as a way to not gain combat skills.

Again, one can start auto-combat and learn certain battle skills as he can watch how AI performs the battle.

**Comment by zenseii** [comment]
Hi, @Alucard648.

The point is that a player won't do this because they will never see the combat screen with instant battles on. The times they do choose to restart they'll just realise that they always get a worse result so they might stick to always accepting the instant result.

**Comment by idshibanov** [comment]
Is it really worth it? We can't police how people are playing the game. IMO if they want to do a quick run and focus on Adventure map - so be it.

When we get to implementing working multiplayer proper auto battles will be the key to save time (mostly early and midgame) and make it reasonable to finishing a game within few hours. If AI won't minimize losses on easy battles nobody will use it.

**Comment by oleg-derevenetz** [comment]
> If AI won't minimize losses on easy battles nobody will use it.

True. Too dumb AI is just useless for a relatively experienced player because instead of saving time, its use only gives unnecessary army losses. It is also useless for a novice player, because, as @Alucard648 said, novice could at least learn some basic tricks from strong AI by watching the battles in auto mode.

So, it is useless for an experienced player and is useless for a novice, what is the target audience of dumb AI then? You might as well just turn the instant battles off, but you can do it now too.

**Comment by LeHerosInconnu** [comment]
Hello everyone,

I think instant battle and auto battle should be as smart as possible, even if the result is better than a human player, otherwise why use them?
The player interested in combats will always try to do better, the player less interested in combats will not be "forced" to do them again, the experienced player will always want to do the most interesting and important combats himself.

**Comment by ihhub** [comment]
As I expected we have many different opinions. I am converting this into discussion.

**Comment by Branikolog** [comment]
Hi, everyone!
After watching some fheroes2 streams I came into conclusion, that even experienced players are satisfied with the outcome of instant battles too often, making a gameplay more adventure map-oriented. Even against rather strong opponents it sometimes performs even better, I could imagine is possible. Any novice player would play worse in the majority of cases, so he never rise skills with such smart AI fighting on his side.
That all goes without saying, that we definitely will improve AI in future even more. I think, current instant battle option starts making game more casual-like.
How many of you played fheroes2, attacked someone on the adventure map with thoughts like: "Oh, there would be some problems with this fight. My army would have suffer some losses"... and was surprised by acceptable outcome after that fight? I think, instant battles were invented for the cases, when you attack a **much** weaker troops, to save you some time. For instance a fight with 2 Black dragons vs 200 peasants is definitely a fight, where instant battle should save players' time.

So in my opinion, we should to keep current instant battle AI plus probably reduce some of his algorithms a little bit to make him behave more straightforward.

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

> After watching some fheroes2 streams I came into conclusion, that even experienced players are satisfied with the outcome of instant battles too often, making a gameplay more adventure map-oriented.

Let the player play the way he wants to play. :)

> So in my opinion, we should to keep current instant battle AI plus probably reduce some of his algorithms a little bit to make him behave more straightforward.

All this will do is cause the player to say:
"I shouldn't have a loss in this combat, this instant/auto combat is stupid and useless." :-1:

**Reply by Branikolog** [reply]
Hi, @LeHerosInconnu !

>"I shouldn't have a loss in this combat, this instant/auto combat is stupid and useless." -1

He could say so, if he'd play any autobattler from the store. :) Battles are 50% of the gameplay and I believe, player should do each battle  against a significant opponent by his own hands. 
I believe, the instant battle shouldn't force player to accept the battle outcome proposed by an AI. I think the instant battle feature shoud provoke a player to win the battle better, than AI.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> Hi, @LeHerosInconnu !
> 
> > "I shouldn't have a loss in this combat, this instant/auto combat is stupid and useless." -1
> 
> He could say so, if he'd play any autobattler from the store. :) Battles are 50% of the gameplay and I believe, player should do each battle against a significant opponent by his own hands. I believe, the instant battle shouldn't force player to accept the battle outcome proposed by an AI. I think the instant battle feature shoud provoke a player to win the battle better, than AI.

I think we're going to waste our time finding the sweet spot for this.
Let's make the most competent AI possible and let the player choose what is right for him (accept the result, redo the combat, throw the computer out of the window, smash his smartphone, etc.).

**Reply by zenseii** [reply]
Personally I would be fine with having instant combat just triggering for the battles against greatly outnumbered armies, both those you refuse to accept and those that try to flee. This could be a compromise between always auto-resolve and never do it. It could be an level of the auto-resolve setting. Just to give an idea like "none/manual", "semi", "full".

EDIT: And before anyone mentions battles against other teams' super weak armies, like fighting an enemy scout, you should think back to the times you've had auto-resolve on for such battles, and then surely you'll remember that you chose to restart the battle because the enemy hero almost 90% of time manages to escape and so you have to restart it manually and try to change the outcome... 

And I am completely fine with having to manually do battles against enemy armies, because they at least involve some tacitic compared to going up against hordes of peasants, which is as easy and as much of a chore to do as taking out the trash.

Actually, taking out the trash at least leaves you with the feeling of having done something useful with your time.

**Comment by Branikolog** [comment]
Bump! 😄 

The case, when experienced players says: "I could finish the battle much better!" - but fails with more losses, happens even more frequently nowadays. :)
I'm still convinced, that current strong battle AI should be applied for human "auto mode" only for "easy" difficulty. I believe for normal and higher difficulties a new AI, which moves more straightforward, should be added. Spell decision could be left as it is, but troop acting should have some trivial logic like: "go and attack the nearest+strongest enemy".

**Reply by oleg-derevenetz** [reply]
> The case, when experienced players says: "I could finish the battle much better!" - but fails with more losses, happens even more frequently nowadays. :)

I believe that this would be the reason for this player to run the battle in auto mode, watch it and maybe learn a few tricks from the AI. Dumb AI is counterproductive in this regard. If a player wants a challenge, he will perform the battles manually (and maybe learn from AI) and if he doesn't want the challenge, then so be it.

**Reply by Branikolog** [reply]
@oleg-derevenetz 
Actually, current AI is the reason for players to reload and accept the instant battle outcome, when he fails playing better. I hardly ever seen a player replaying the battle just to watch, how was it possible. Here I'm talking about experienced players. 
For novice players, who study the game on "easy" difficulty, the instant battle AI would still perform its best, so they can watch all the tricks. :)
The thing bothers me, that players tend not to use manual battles at all now. I've seen a few streams some time ago, when the player hardly ever used manual battle mode for the whole campaign scenario. Instant battle outcomes were so good, so he didn't want to replay. In some cases even I was wondering, how could AI actually won that particular battle.

**Reply by oleg-derevenetz** [reply]
> I've seen a few streams some time ago, when the player hardly ever used manual battle mode for the whole campaign scenario.

I occasionally watch fheroes2 videos from the guy with the nickname "KhelbensCoin" and he always plays with the instant battle turned off to have a challenge. So if player wants a challenge, there is still no problem to have it. If other players do not want to have a challenge, or want to concentrate on strategical aspects, or want to use the auto battle to learn something, then this should be their own choice IMHO.

**Reply by Branikolog** [reply]
Hi, @oleg-derevenetz .
I apologize if I'm bothering you too much with this topic. :)

"KhelbensCoin" is a nice old-school H2 player, who used to play the original game without instant battles through all his life. He also prefers manual battles, I believe, because it makes streaming or lets-play videos more exciting. :)

Fheroes2 AI is inevitably becoming stronger and stronger. And I even find myself skipping tough battles sometimes, because AI somehow managed to win with minimal losses. I'm afraid, some day we will no longer need manual battles at all. :(

> If other players do not want to have a challenge, or want to concentrate on strategical aspects

Okay. If come players don't like a strategical aspect and love only battles, should we allow AI to handle all strategy on the adventure map (as it is implemented for debug mode) and allow player to control troops in battles only then? :D (Actually, we've had a proposal some time ago from our community, to make this mode available for master build, so human can just sit and relax watching cold machines demolishing each other on the fields of Enroth. 😃 )

Battles are an important part of playing experience, and I believe a game should always motivate players to take part in battles (the hardest ones). I am personally find myself really satisfied and even happy, when the instant battle showed some essential losses, but I managed to make them minimal. 
Instant battle feature is designed to be a tool "to skip easy battles". But not to completely substitute player in all battles.

I'm not proposing making AI completely dumb and useless. Hero shouldn't loose dragons in battles against peasants for sure. :) I expect instant battle AI to perform +/- decent. An assistant, that can easily handle battles, when hero army is ~3-4 times stronger, than the opponent's. Similar, how it was in H3.
In the distant future, with implementing multiplayer mode, lots of online battles will be held with a timing restrictions. So players should compete conducting fast battles against neutral monsters. Using mighty fheroes2 battle AI (in instant or auto battles) will definitely cut this part of challenge between players. 😿 

The battle logic should be simplified into somewhat straightforward. AI should be in attacking state always, without maneuvering or abusing special abilities of troops. Such implementation could also help a player to understand, what outcome to expect, in a case he chooses the offensive tactics. (Not, that this change should be applied only for the AI, that controls human's troops. The opposing AI should perform his best!)

**Comment by LeHerosInconnu** [comment]
Hello @Branikolog,

> The case, when experienced players says: "I could finish the battle much better!" - but fails with more losses, happens even more frequently nowadays. :)

So what's the problem then?
The player thought he could do better, but he couldn't by replaying the combat.

> Actually, current AI is the reason for players to reload and accept the instant battle outcome, when he fails playing better. I hardly ever seen a player replaying the battle just to watch, how was it possible. Here I'm talking about experienced players. For novice players, who study the game on "easy" difficulty, the instant battle AI would still perform its best, so they can watch all the tricks. :)
The thing bothers me, that players tend not to use manual battles at all now. I've seen a few streams some time ago, when the player hardly ever used manual battle mode for the whole campaign scenario. Instant battle outcomes were so good, so he didn't want to replay. In some cases even I was wondering, how could AI actually won that particular battle.

Let the player play the game in the way that suits him best.
It should be a pleasure to play, not a constraint.

**Comment by Granfalegion** [comment]
I do feel like the goal of improving the vanilla AI may have already been achieved, and that further improvements are to the detriment of the experience.

My understanding was that the new AI was meant to replicate the behavior of a novice player: better than a blind full-speed rush towards enemy lines, but not so intelligent that it fails to make mistakes.  It strikes me that an AI that plays better than "professionals" isn't meeting that need.

**Reply by oleg-derevenetz** [reply]
> My understanding was that the new AI was meant to replicate the behavior of a novice player: better than a blind full-speed rush towards enemy lines, but not so intelligent that it fails to make mistakes. It strikes me that an AI that plays better than "professionals" isn't meeting that need.

There is another opinion - AI development should gradually move towards its further improvement (taking into account the limitations in terms of comfortable decision-making speed, especially on weak computers or handheld devices). Playing against a beginner is pretty boring. Also, I don't see a big tragedy in the fact that AI can occasionally play better than an experienced player. On the contrary, it can be interesting for a player to compete with AI and maybe even learn something from it by replaying the combat in AUTO mode and by observing the actions taken by AI to get a better result than the human player had.

**Reply by ihhub** [reply]
Hi @Granfalegion , we will try to easy the AI behavior (not to make it stupid, just to restrict in doing things) for lower difficulties but Hard and above should be a challenge for any player based on the name of difficulties. As of now the AI persists a little challenge on Impossible difficulty for very experienced players and some like @oleg-derevenetz would like to have a good challenge on it. So we are working towards smarter version of it for higher difficulties but will try to restrict in on lower difficulties.

From what I observe the AI plays worse in battles in 99% cases for me :)

**Reply by Granfalegion** [reply]
A sliding difficulty sounds great.  An ability to control how what sort of AI your enemy has sounds great.

In those situations, would the proposal also alter how your own auto-control reacts?  It would be weird for your own troops to perform better on Impossible than they do on Normal (even if the expectation is that the opposition does as well).

**Comment by ihhub** [comment]
Since the opening of this issue and discussion and looking at everyone's comments and especially @oleg-derevenetz I have doubts that this is a good idea...

**Comment by minuspinus** [comment]
Maybe we should introduce a fine for Autofight. 
You can't get out of an autofight without losses.

Kind of like whirlpools on water.

**Comment by modo-lv** [comment]
I like @zenseii's idea of "semi" instant-battle option that would make it trigger only on greatly outnumbered enemies. That way the AI doesn't have to be specially adjusted, and players won't be tempted into skipping interesting fights, while still saving time on trivial ones. I would definitely prefer playing with that option.

As for the idea that novice players could use auto battle to learn — that's something I hadn't even thought of before, and I really like the idea. Maybe it's worth adding some kind of an "watch battle" button (that just restarts the fight with auto battle switched on) for the instant battle screen? It would remind players that instant battle and auto battle are the same thing and they can actually see what the AI did differently if the result was better than the player could do.
